File: server/app/src/main/java/org/MercuryWorkshop/moltenserver/MainActivity.java
package org.MercuryWorkshop.moltenserver;

import java.io.BufferedInputStream;
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.io.Reader;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.ServerSocket;
import java.net.Socket;
import java.net.URL;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.security.Permissions;
import java.util.Base64;
import java.util.Iterator;
import java.util.Scanner;

import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import android.Manifest;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.util.Log;
import android.widget.Button;
import android.widget.TextView;
import org.json.*;

import javax.net.ssl.HttpsURLConnection;

import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import okhttp3.internal.http.HttpHeaders;
import okhttp3.internal.http2.Header;

public class MainActivity extends AppCompatActivity {

    private Button buttonStartReceiving;
    private Button buttonStopReceiving;
    private TextView textViewDataFromClient;
    final Handler handler = new Handler();
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        textViewDataFromClient = (TextView) findViewById(R.id.textViewDataFromClient);

        if (ContextCompat.checkSelfPermission(this, Manifest.permission.MANAGE_EXTERNAL_STORAGE)
                != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this,
                    new String[]{Manifest.permission.MANAGE_EXTERNAL_STORAGE},
                    1);
            Log.e("MoltenServer","jsakd");

        }else{
            Log.e("MoltenServer","ASDJKASDL");
        }
//        Thread thread = new Thread(new Runnable() {
//
//            OkHttpClient client = new OkHttpClient();
//            URL url = null;
//            @Override
//            public void run() {
//                try {
//                    url = new URL("https://www.google.com/images/branding/googlelogo/2x/googlelogo_color_272x92dp.png");
//                } catch (
//                        MalformedURLException e) {
//                    throw new RuntimeException(e);
//                }
//
//                Request req = new Request.Builder().url(url).build();
//                try (
//                        Response httpResp = client.newCall(req).execute()) {
//
//
//                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
//                        byte[] b = httpResp.body().source().readByteArray();
////                        byte[] body = httpResp.body().bytes();
////                        body.toString();
////                        String read = httpResp.body().source().readUtf8();
////
////                        ByteArrayOutputStream bOut = new ByteArrayOutputStream();
////
////                        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(new FileOutputStream("/storage/emulated/0/Documents/fuck.png"), "utf8"));
////
////                        bw.write(Base64.getMimeEncoder().encodeToString(b));
//                        bw.close();
//                    }
//                } catch (
//                        IOException e) {
//                    throw new RuntimeException(e);
//                }
//            }
//        });
//        thread.start();
        startServerSocket();
    }
    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == 1) {
            if (grantResults.length > 0
                    && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                Log.d("MoltenServer","thranted??/");
                // Permission granted, do your work
            } else {
                Log.d("MoltenServer","asdjaskldjklas");
//                Log.d("MoltenServer",permissions);
            }
        }
    }

    private void listenToClient(Socket sock) {
        Thread thread = new Thread(new Runnable() {
            private boolean end = false;
            private String stringData = null;
            OkHttpClient client = new OkHttpClient().newBuilder()
                    .followRedirects(false)
                    .followSslRedirects(false)
                    .build();

            @Override
            public void run() {
                try {


                    BufferedReader input = new BufferedReader(new InputStreamReader(sock.getInputStream()));
                    PrintWriter output = new PrintWriter(sock.getOutputStream());
                    while (!end) {
//                        output.println("CONNECTION");
                        stringData = input.readLine();
                        if (stringData == null) {
                            end = true;
                            return;
                        }
                        try {
                            JSONObject request = new JSONObject(stringData);
                            JSONObject reqHeaders = request.getJSONObject("headers");
                            URL url = new URL(request.getString("url"));
                            Request.Builder builder = new Request.Builder().url(url);


                            if (!request.getString("method").equalsIgnoreCase("get")) {
                                Log.d("MoltenServer",request.getString("method"));
                                builder.method(request.getString("method"), RequestBody.create(MediaType.parse(reqHeaders.getString("Content-Type")), request.getString("body")));
                            }
                            for (Iterator<String> it = reqHeaders.keys(); it.hasNext(); ) {
                                String key = it.next();
                                if (!key.equals("Authorization")) {
                                    builder.header(key, reqHeaders.getString(key));
                                }
                            }


                            Request httpreq = builder.build();
                            try (Response httpResp = client.newCall(httpreq).execute()) {


                                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                                    byte[] body = httpResp.body().source().readByteArray();

                                    JSONObject resp = new JSONObject();
                                    resp.put("id", request.getString("id"));
                                    resp.put("data", Base64.getMimeEncoder().encodeToString(body));


                                    JSONObject headers = new JSONObject();

                                    for (Iterator<String> it = httpResp.headers().toMultimap().keySet().iterator(); it.hasNext(); ) {
                                        String key = it.next();
                                        headers.put(key,httpResp.header(key));
                                    }
                                    resp.put("status",httpResp.code());
                                    resp.put("headers",headers);
                                    output.println(resp.toString() + "\004");

                                }
                            }
                        }catch (Exception e){
                            Log.d("MoltenServer",e.toString());
                        }

                        Log.e("MoltenServer",stringData += ":EOL");
                        output.flush();
//                        android.os.Process.killProcess(android.os.Process.myPid());


//                        try {
//                            Thread.sleep(1000);
//                        } catch (InterruptedException e) {
//                            e.printStackTrace();
//                        }

                        //output.
                        updateUI(stringData);
                        if (stringData.equalsIgnoreCase("STOP")) {
                            end = true;
                            output.close();
                            try {
                                sock.close();
                            } catch (IOException e) {
                                throw new RuntimeException(e);
                            }
                            break;
                        }

                    }
                    output.close();
                    sock.close();
                }catch (IOException e){
                    throw new RuntimeException(e);
                }
            }
        });
        thread.start();

    }
    private void startServerSocket() {

        Thread thread = new Thread(new Runnable() {

            private String stringData = null;

            @Override
            public void run() {

                try {

                    ServerSocket ss = new ServerSocket(1234);

                    while (true) {
                        //Server is waiting for client here, if needed
                        Socket s = ss.accept();
                        listenToClient(s);
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }

        });
        thread.start();
    }

    private void updateUI(final String stringData) {

        handler.post(new Runnable() {
            @Override
            public void run() {

                String s = textViewDataFromClient.getText().toString();
                if (stringData.trim().length() != 0)
                    textViewDataFromClient.setText(s + "\n" + "From Client : " + stringData);
            }
        });
    }
}

File: server/app/src/main/java/org/MercuryWorkshop/moltenserver/MainActivity.kt
package org.MercuryWorkshop.moltenserver

import android.Manifest
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.TextView
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import okhttp3.MediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import org.json.JSONObject
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.io.PrintWriter
import java.net.ServerSocket
import java.net.Socket
import java.net.URL
import java.util.*

class MainActivity() : AppCompatActivity() {
    private val buttonStartReceiving: Button? = null
    private val buttonStopReceiving: Button? = null
    private var textViewDataFromClient: TextView? = null
    val handler = Handler()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        textViewDataFromClient = findViewById<View>(R.id.textViewDataFromClient) as TextView
        if ((ContextCompat.checkSelfPermission(this, Manifest.permission.MANAGE_EXTERNAL_STORAGE)
                    != PackageManager.PERMISSION_GRANTED)
        ) {
            ActivityCompat.requestPermissions(
                this, arrayOf(Manifest.permission.MANAGE_EXTERNAL_STORAGE),
                1
            )
            Log.e("MoltenServer", "jsakd")
        } else {
            Log.e("MoltenServer", "ASDJKASDL")
        }
        //        Thread thread = new Thread(new Runnable() {
//
//            OkHttpClient client = new OkHttpClient();
//            URL url = null;
//            @Override
//            public void run() {
//                try {
//                    url = new URL("https://www.google.com/images/branding/googlelogo/2x/googlelogo_color_272x92dp.png");
//                } catch (
//                        MalformedURLException e) {
//                    throw new RuntimeException(e);
//                }
//
//                Request req = new Request.Builder().url(url).build();
//                try (
//                        Response httpResp = client.newCall(req).execute()) {
//
//
//                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
//                        byte[] b = httpResp.body().source().readByteArray();
////                        byte[] body = httpResp.body().bytes();
////                        body.toString();
////                        String read = httpResp.body().source().readUtf8();
////
////                        ByteArrayOutputStream bOut = new ByteArrayOutputStream();
////
////                        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(new FileOutputStream("/storage/emulated/0/Documents/fuck.png"), "utf8"));
////
////                        bw.write(Base64.getMimeEncoder().encodeToString(b));
//                        bw.close();
//                    }
//                } catch (
//                        IOException e) {
//                    throw new RuntimeException(e);
//                }
//            }
//        });
//        thread.start();
        startServerSocket()
    }

    override fun onRequestPermissionsResult(
        requestCode: Int, permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == 1) {
            if ((grantResults.size > 0
                        && grantResults[0] == PackageManager.PERMISSION_GRANTED)
            ) {
                Log.d("MoltenServer", "thranted??/")
                // Permission granted, do your work
            } else {
                Log.d("MoltenServer", "asdjaskldjklas")
                //                Log.d("MoltenServer",permissions);
            }
        }
    }

    private fun listenToClient(sock: Socket) {
        val thread = Thread(object : Runnable {
            private var end = false
            private var stringData: String? = null
            var client = OkHttpClient().newBuilder()
                .followRedirects(false)
                .followSslRedirects(false)
                .build()

            @RequiresApi(Build.VERSION_CODES.O)
            override fun run() {
                try {
                    val input = BufferedReader(InputStreamReader(sock.getInputStream()))
                    val output = PrintWriter(sock.getOutputStream())
                    while (!end) {
//                        output.println("CONNECTION");
                        stringData = input.readLine()
                        if (stringData == null) {
                            end = true
                            return
                        }
                        try {
                            val request = JSONObject(stringData)
                            val reqHeaders = request.getJSONObject("headers")
                            val url = URL(request.getString("url"))
                            val builder = Request.Builder().url(url)
                            if (!request.getString("method").equals("get", ignoreCase = true)) {
                                Log.d("MoltenServer", request.getString("method"))
                                builder.method(
                                    request.getString("method"), RequestBody.create(
                                        MediaType.parse(reqHeaders.getString("Content-Type")),
                                        request.getString("body")
                                    )
                                )
                            }
                            val it = reqHeaders.keys()
                            while (it.hasNext()) {
                                val key = it.next()
                                if (key != "Authorization") {
                                    builder.header(key, reqHeaders.getString(key))
                                }
                            }
                            val httpreq = builder.build()
                            try {
                                client.newCall(httpreq).execute().use { httpResp ->
                                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                                        val body: ByteArray =
                                            httpResp.body()!!.source().readByteArray()
                                        val resp: JSONObject = JSONObject()
                                        resp.put("id", request.getString("id"))
                                        resp.put(
                                            "data",
                                            Base64.getMimeEncoder().encodeToString(body)
                                        )
                                        val headers: JSONObject = JSONObject()
                                        val it: Iterator<String> =
                                            httpResp.headers().toMultimap().keys.iterator()
                                        while (it.hasNext()) {
                                            val key: String = it.next()
                                            headers.put(key, httpResp.header(key))
                                        }
                                        resp.put("status", httpResp.code())
                                        resp.put("headers", headers)
                                        output.println(resp.toString() + "\u0004")
                                    }
                                }
                            }catch (e: Exception){
                                val resp: JSONObject = JSONObject()
                                resp.put("id", request.getString("id"))
                                resp.put(
                                    "error",
                                    Base64.getMimeEncoder().encodeToString(e.toString().toByteArray())
                                )
                                output.println(resp.toString() + "\u0004")
                            }
                        } catch (e: Exception) {
                            Log.d("MoltenServer", e.toString())
                        }
                        Log.e("MoltenServer", (":EOL".let { stringData += it; stringData })!!)
                        output.flush()
                        //                        android.os.Process.killProcess(android.os.Process.myPid());


//                        try {
//                            Thread.sleep(1000);
//                        } catch (InterruptedException e) {
//                            e.printStackTrace();
//                        }

                        //output.
                        updateUI(stringData!!)
                        if (stringData.equals("STOP", ignoreCase = true)) {
                            end = true
                            output.close()
                            try {
                                sock.close()
                            } catch (e: IOException) {
                                throw RuntimeException(e)
                            }
                            break
                        }
                    }
                    output.close()
                    sock.close()
                } catch (e: IOException) {
                    throw RuntimeException(e)
                }
            }
        })
        thread.start()
    }

    private fun startServerSocket() {
        val thread = Thread(object : Runnable {
            private val stringData: String? = null
            override fun run() {
                try {
                    val ss = ServerSocket(1234)
                    while (true) {
                        //Server is waiting for client here, if needed
                        val s = ss.accept()
                        listenToClient(s)
                    }
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
        })
        thread.start()
    }

    private fun updateUI(stringData: String) {
        handler.post(Runnable {
            val s = textViewDataFromClient!!.text.toString()
            if (stringData.trim { it <= ' ' }.length != 0) textViewDataFromClient!!.text =
                "$s\nFrom Client : $stringData"
        })
    }
}